blocklylib-core/src/main/java/com/google/blockly/android/codegen/CodeGeneratorManager.java

```diff
@@ -43,54 +43,41 @@ public void onServiceDisconnected(ComponentName arg0) {
         };
     }
 
-    public void onResume() {
-        Intent intent = new Intent(mContext, CodeGeneratorService.class);
-        mContext.bindService(intent, mCodeGenerationConnection, Context.BIND_AUTO_CREATE);
-    }
-
+    /**
+     * Unbind the underlying service (if it is bound).
+     */
     public void onPause() {
-        mContext.unbindService(mCodeGenerationConnection);
+        if (isBound()) {
```

Comment:
In order to deal with race conditions, including the user leaving and re-entering while a bind request is in progress, you'll want to keep track of the desired state.
- In onPause you should also set a mResumed flag to false (and set it to true in onResume). 
- In onServiceConnected check if (mResumed). If not, disconnected immediately.
- In onResume you should clear out any pending requests since you won't reconnect until a new request comes in.